api/v1alpha1/gateway_types.go

```diff
@@ -54,78 +57,200 @@ type GatewayList struct {
+	// this in the associated entry in GatewayStatus.Listeners. If no
+	// addresses are specified, the GatewayClass may schedule the Gateway
+	// in an implementation-defined manner, providing it with an appropriate
+	// set of addresses.
 	//
-	// Support: Core
+	// Support:
 	//
-	Routes RouteBindingSelector `json:"routes" protobuf:"bytes,3,opt,name=routes"`
+	// +optional
+	Addresses []ListenerAddress `json:"endpoints" protobuf:"bytes,3,rep,name=endpoints"`
 }
 
+// ProtocolType defines the application protocol accepted by a Listener.
+type ProtocolType string
+
+const (
+	// HTTPProtocolType accepts cleartext HTTP/1.1 sessions over TCP.
+	HTTPProtocolType ProtocolType = "HTTP"
+
+	// HTTPSProtocolType accepts HTTP/1.1 or HTTP/2 sessions over TLS.
+	HTTPSProtocolType ProtocolType = "HTTPS"
+
+	// TLSProtocolType accepts TLS sessions over TCP.
+	TLSProtocolType ProtocolType = "TLS"
+
+	// TCPProtocolType accepts TCP sessions.
+	TCPProtocolType ProtocolType = "TCP"
+)
+
+// HostnameMatchType specifies the types of matches that are valid
+// for host names.
+type HostnameMatchType string
+
 const (
-	// HTTPProtocol constant.
-	HTTPProtocol = "HTTP"
-	// HTTPSProtocol constant.
-	HTTPSProtocol = "HTTPS"
+	// HostnameMatchDomain specifies that the host name provided by the
+	// client should be matched against a DNS domain value. The domain
+	// match should follow the wildcard certificate matching rules in
+	// [RFC 6125](https://tools.ietf.org/html/rfc6125#section-6.4.3).
+	HostnameMatchDomain HostnameMatchType = "Domain"
+
+	// HostnameMatchExact specifies that the host name provided by the
+	// client must exactly match the specified value.
+	HostnameMatchExact HostnameMatchType = "Exact"
 )
 
-// Listener defines a
+// HostnameMatch specifies how a Listener should match the host name from a
+// client request. Depending on the incoming protocol, the match must apply to
+// names provided by the client at both the TLS and the HTTP protocol layers.
+type HostnameMatch struct {
+	// Match specifies how the host name provided by the client should be
+	// matched against the given value.
+	//
+	// +optional
+	// +kubebuilder:validation:Enum=Domain;Exact
+	// +kubebuilder:default=Exact
+	Match HostnameMatchType `json:"match" protobuf:"bytes,1,name=match"`
+
+	// Name contains the name to match against. This value must be a
+	// fully qualified host or domain name as defined by
+	// [RFC 3986](https://tools.ietf.org/html/rfc3986#section-3.2.2).
+	// Note the following deviations from the "host" part of the URI as defined in the RFC:
+	//
+	// 1. IPs are not allowed.
+	// 2. The `:` delimiter is not respected because ports are not allowed.
+	//
+	// +required
+	// +kubebuilder:validation:MinLength=1
+	Name string `json:"value" protobuf:"bytes,2,name=value"`
+}
+
+// Listener embodies the concept of a logical endpoint where a
+// Gateway can accept network connections.
+//
+// The GatewayClass may collapse Listener definitions into single
```

Comment:
> Is this "May" or "Must"?

Currently MAY, because I didn't want to require implementations to do this. Maybe some gateway classes just don't implement layer 7 matching but should otherwise be conformant.

> Hard to see how a Gateway with a PortConflict can be implemented?

Not sure whether I'm interpreting your question correctly, but if you add 2 Listeners with matching Port fields and TCP protocol, that would be a PortConflict (since they are not collapsible).

> Can we use the term "compatible Listener" instead of collapsible?

Previously I used "compatible", but it was less specific than using "collapsible" as a term of art. Not sure what the right approach is here; using a more generic word makes it harder to mean something extremely specific.

> Should we just say that within a given Gateway, they must all be compatible?

Yes, but what language should we use to describe what compatibility means?

I'll try another pass at this in terms of compatibility and see whether that works out ...